### PR TITLE
task completed

### PR DESCRIPTION
**Pull request by ArthurKrepko**
[DEMO LINK](https://ArthurKrepko.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/types/User.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export interface User {
+  id: number;
```
Comment (Andr1yK) [inline src/types/User.ts:2]: Consider adding more properties to the User interface if needed. 🤔

File: src/types/User.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+  name: string;
```
Comment (Andr1yK) [inline src/types/User.ts:3]: Consider using a more descriptive name for the `id` property, like `userId`. 🤔

File: src/types/Todo.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface Todo {
+  id: number;
```
Comment (Andr1yK) [inline src/types/Todo.ts:2]: Consider adding JSDoc comments to describe the purpose of the interface and its properties. 📝

Comment (Andr1yK) [inline src/types/Todo.ts:2]: Consider renaming `Todo` interface to `ITodo` to follow common naming conventions for interfaces. 🤔

File: src/components/UserInfo/UserInfo.tsx [hunk 4]
```diff
@@ -1 +1,17 @@
...
+
+type Props = {
+  user: User;
+};
+
+export const UserInfo: React.FC<Props> = ({ user }) => {
+  const { email, name } = user;
+
+  return (
+    <a className="UserInfo" href={`mailto:${email}`}>
```
Comment (Andr1yK) [inline src/components/UserInfo/UserInfo.tsx:13]: Consider adding `rel='noopener noreferrer'` to the anchor tag for better security. 🔒

File: src/components/TodoList/TodoList.tsx [hunk 5]
```diff
@@ -1 +1,17 @@
-export const TodoList = () => {};
```
Comment (Andr1yK) [inline src/components/TodoList/TodoList.tsx:1]: Consider removing the unused export statement. 🤔

File: src/components/TodoList/TodoList.tsx [hunk 6]
```diff
@@ -1 +1,17 @@
...
+import React from 'react';
+import { FullTodo } from '../../types/FullTodo';
+import { TodoInfo } from '../TodoInfo';
```
Comment (Andr1yK) [inline src/components/TodoList/TodoList.tsx:3]: Consider importing `FullTodo` type from a separate file to avoid duplication. 📑

File: src/components/UserInfo/UserInfo.tsx [hunk 7]
```diff
@@ -1 +1,17 @@
-export const UserInfo = () => {};
+import React from 'react';
+import { User } from '../../types/User';
+import './UserInfo.scss';
```
Comment (Andr1yK) [inline src/components/UserInfo/UserInfo.tsx:3]: Consider organizing imports in alphabetical order for better readability. 📚

File: src/components/UserInfo/UserInfo.tsx [hunk 8]
```diff
@@ -1 +1,17 @@
-export const UserInfo = () => {};
```
Comment (Andr1yK) [inline src/components/UserInfo/UserInfo.tsx:1]: Consider removing unused export statement. 🗑️

File: src/components/TodoList/TodoList.tsx [hunk 9]
```diff
@@ -1 +1,17 @@
...
+
+type Props = {
+  todos: FullTodo[];
+};
+
+export const TodoList: React.FC<Props> = ({ todos }) => {
+  return (
+    <section className="TodoList">
```
Comment (Andr1yK) [inline src/components/TodoList/TodoList.tsx:11]: Consider using a more descriptive class name instead of `TodoList`. 📝

File: src/components/TodoInfo/TodoInfo.tsx [hunk 10]
```diff
@@ -1 +1,37 @@
-export const TodoInfo = () => {};
+import classNames from 'classnames';
```
Comment (Andr1yK) [inline src/components/TodoInfo/TodoInfo.tsx:1]: Consider alphabetizing imports for better readability and consistency. 📚

File: src/components/TodoInfo/TodoInfo.tsx [hunk 11]
```diff
@@ -1 +1,37 @@
...
+import React from 'react';
+import { FullTodo } from '../../types/FullTodo';
+import { UserInfo } from '../UserInfo';
+import './TodoInfo.scss';
```
Comment (Andr1yK) [inline src/components/TodoInfo/TodoInfo.tsx:5]: Consider moving the CSS import to a separate file or module for better separation of concerns. 💅

File: src/components/TodoInfo/TodoInfo.tsx [hunk 12]
```diff
@@ -1 +1,37 @@
...
+    user,
+    id,
+    completed,
+  } = todo;
+
+  return (
+    <article
+      data-id={id}
+      className={
+        classNames('TodoInfo', { 'TodoInfo--completed': completed })
```
Comment (Andr1yK) [inline src/components/TodoInfo/TodoInfo.tsx:23]: Consider using a more descriptive class name for better readability and understanding. 🤔

File: src/types/FullTodo.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import { Todo } from './Todo';
```
Comment (Andr1yK) [inline src/types/FullTodo.ts:1]: Consider using relative path instead of absolute path for better maintainability. 🛣️

File: src/types/FullTodo.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
...
+import { User } from './User';
+
+export interface FullTodo extends Todo {
+  user: User | null;
```
Comment (Andr1yK) [inline src/types/FullTodo.ts:5]: Consider using a more descriptive name for the property `user`. For example, `assignedUser` or `assignedTo`. 🤔

File: src/types/FullTodo.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+import { Todo } from './Todo';
+import { User } from './User';
+
+export interface FullTodo extends Todo {
```
Comment (Andr1yK) [inline src/types/FullTodo.ts:4]: Interface `FullTodo` extends `Todo` but it is not clear what additional properties it adds. Consider renaming the interface to better reflect its purpose. 🤔

File: src/components/TodoInfo/TodoInfo.scss [hunk 16]
```diff
@@ -0,0 +1,18 @@
+.TodoInfo {
+  width: max-content;
+  padding: 8px;
+  margin: 12px 0;
+  border: 1px solid #000;
```
Comment (Andr1yK) [inline src/components/TodoInfo/TodoInfo.scss:5]: Consider using variables for border and background colors to avoid hardcoding. 💡

File: src/components/TodoInfo/TodoInfo.scss [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+  border-radius: 8px;
+  background-color: antiquewhite;
+
+  &__title {
+    margin: 4px 0;
+    font-size: inherit;
+    color: #f00;
+  }
+
+  &--completed &__title {
```
Comment (Andr1yK) [inline src/components/TodoInfo/TodoInfo.scss:15]: Consider using BEM naming convention for better readability and maintainability. For example, `.todo-info-container--completed__title`. 🤓

File: src/components/TodoInfo/TodoInfo.scss [hunk 18]
```diff
@@ -0,0 +1,18 @@
+.TodoInfo {
+  width: max-content;
```
Comment (Andr1yK) [inline src/components/TodoInfo/TodoInfo.scss:2]: Consider using more descriptive class name instead of `.TodoInfo`. For example, `.todo-info-container` or `.todo-info-wrapper`. 🤔

File: src/components/UserInfo/UserInfo.scss [hunk 19]
```diff
@@ -0,0 +1,3 @@
+.UserInfo {
+  font-style: italic;
```
Comment (Andr1yK) [inline src/components/UserInfo/UserInfo.scss:2]: Consider using a more specific CSS selector to avoid unintended styling of other elements. For example, `.user-info-container .user-info` or `.user-info-wrapper > .user-info`. 🎯

Comment (Andr1yK) [inline src/components/UserInfo/UserInfo.scss:2]: Consider using a more descriptive class name than `.UserInfo`. For example, `.user-info-container` or `.user-info-wrapper`. 🤔

File: src/components/UserInfo/UserInfo.scss [hunk 20]
```diff
@@ -0,0 +1,3 @@
...
+}
```
Comment (Andr1yK) [inline src/components/UserInfo/UserInfo.scss:3]: Consider using a more descriptive property name than `font-style`. For example, `font-family` or `font-weight`. 🖋️

File: src/App.scss [hunk 21]
```diff
@@ -14,26 +14,3 @@ button {
 .error {
   color: red;
 }
-
-.TodoInfo {
```
Comment (Andr1yK) [inline src/App.scss:18]: Consider removing `.TodoInfo` class as it is no longer used. 🧹

File: src/App.scss [hunk 22]
```diff
@@ -14,26 +14,3 @@ button {
...
-    font-size: inherit;
-    color: #f00;
-  }
-
-  &--completed &__title {
-    color: #080;
-  }
-}
-
-.UserInfo {
```
Comment (Andr1yK) [inline src/App.scss:37]: Consider using a more descriptive class name than `.UserInfo`. 🤔

File: src/App.tsx [hunk 23]
```diff
@@ -1,61 +1,124 @@
...
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { User } from './types/User';
+import { FullTodo } from './types/FullTodo';
+import { TodoList } from './components/TodoList';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUserById(userId: number): User | null {
```
Comment (Andr1yK) [inline src/App.tsx:10]: Consider renaming `getUserById` to `findUserById` to better describe what the function does. 🤔

File: src/App.tsx [hunk 24]
```diff
@@ -1,61 +1,124 @@
...
+      user: getUserById(selectedUserId),
+    };
+
+    setTodos([...todos, newTodo]);
+  };
+
+  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (newTitle && selectedUserId) {
```
Comment (Andr1yK) [inline src/App.tsx:49]: Consider using optional chaining to avoid errors when `newTitle` or `selectedUserId` is null or undefined. 🚫

File: src/App.tsx [hunk 25]
```diff
@@ -1,61 +1,124 @@
+import React, { useEffect, useState } from 'react';
 import './App.scss';
 
-// import usersFromServer from './api/users';
```
Comment (Andr1yK) [inline src/App.tsx:3]: Consider removing commented out code to improve readability and maintainability. 🧹

File: src/App.tsx [hunk 26]
```diff
@@ -1,61 +1,124 @@
...
+    return copy[0].id + 1;
+  };
+
+  const addNewTodo = (title: string, userId: number) => {
+    const newTodo: FullTodo = {
+      id: getMaxId(todos),
+      title,
+      completed: false,
+      userId,
+      user: getUserById(selectedUserId),
```
Comment (Andr1yK) [inline src/App.tsx:40]: Consider using `todo.userId` instead of `selectedUserId` to avoid confusion and improve readability. 🤔

File: src/App.tsx [hunk 27]
```diff
@@ -1,61 +1,124 @@
...
+    event.preventDefault();
+
+    if (newTitle && selectedUserId) {
+      addNewTodo(newTitle, selectedUserId);
+      setNewTitle('');
+      setNewUserId(0);
+    }
+  };
+
+  useEffect(() => {
```
Comment (Andr1yK) [inline src/App.tsx:56]: Consider using a custom hook to handle form validation instead of using `useEffect` to improve readability and maintainability. 🎣

File: src/App.tsx [hunk 28]
```diff
@@ -1,61 +1,124 @@
...
+
+function getUserById(userId: number): User | null {
+  const foundUser = usersFromServer.find(user => user.id === userId);
+
+  return foundUser || null;
+}
+
+const todosWithUsers: FullTodo[] = todosFromServer.map(todo => ({
+  ...todo,
+  user: getUserById(todo.userId),
```
Comment (Andr1yK) [inline src/App.tsx:18]: Consider using object spread operator instead of `Object.assign` to improve readability. 📖

File: src/App.tsx [hunk 29]
```diff
@@ -1,61 +1,124 @@
...
+}));
+
+export const App: React.FC = () => {
+  const [todos, setTodos] = useState<FullTodo[]>(todosWithUsers);
+  const [newTitle, setNewTitle] = useState('');
```
Comment (Andr1yK) [inline src/App.tsx:23]:
```suggestion
  const [title, setTitle] = useState('');
```

File: src/App.tsx [hunk 30]
```diff
@@ -1,61 +1,124 @@
...
+  const [selectedUserId, setNewUserId] = useState(0);
```
Comment (Andr1yK) [inline src/App.tsx:24]: In React, when using the `useState` hook to manage state, it’s a common convention to name the state variable and its corresponding setter function using a specific naming template: `[variableName, “set” + variableName]`.

For example, in the following code:

```js
const [isReversed, setReverse] = useState(false);
```

`isReversed` is the state variable, and setReverse is the setter function for that variable.

Using this naming convention has several advantages:

- **Readability**: It makes it clear that the setReverse function is specifically for setting the `isReversed` state variable, which can improve the readability of the code.

- **Consistency**: Using a consistent naming convention makes it easier for other developers to understand the codebase and reduces the cognitive load of working with it.

- **Autocompletion**: Many code editors and IDEs will suggest the correct setter function name based on the state variable name, which can make it faster to write code.

Here’s an example of how we can use this naming convention with multiple state variables:

```js
const [isReversed, setIsReversed] = useState(false);
const [selectedItem, setSelectedItem] = useState(null);
const [count, setCount] = useState(0);
```

In summary, naming your React state variables and setter functions using the `[variableName, “set” + variableName]` template can improve the readability, consistency, and autocompletion of your code.

File: src/App.tsx [hunk 31]
```diff
@@ -1,61 +1,124 @@
...
+  const [hasTitleError, setHasTitleError] = useState(false);
+  const [hasUserError, setHasUserError] = useState(false);
+
+  const getMaxId = (todosArray: FullTodo[]): number => {
+    const copy = [...todosArray].sort((prev, curr) => curr.id - prev.id);
+
+    return copy[0].id + 1;
+  };
```
Comment (Andr1yK) [inline src/App.tsx:32]:
```suggestion
  const getNextId = (todosArray: FullTodo[]): number => {
    const copy = [...todosArray].sort((prev, curr) => curr.id - prev.id);

    return copy[0].id + 1;
  };
```

File: src/App.tsx [hunk 32]
```diff
@@ -1,61 +1,124 @@
...
+
+export const App: React.FC = () => {
+  const [todos, setTodos] = useState<FullTodo[]>(todosWithUsers);
+  const [newTitle, setNewTitle] = useState('');
+  const [selectedUserId, setNewUserId] = useState(0);
+  const [hasTitleError, setHasTitleError] = useState(false);
+  const [hasUserError, setHasUserError] = useState(false);
+
+  const getMaxId = (todosArray: FullTodo[]): number => {
+    const copy = [...todosArray].sort((prev, curr) => curr.id - prev.id);
```
Comment (Andr1yK) [inline src/App.tsx:29]: let's use reduce or map + math.max to reduce complexity

File: src/App.tsx [hunk 33]
```diff
@@ -1,61 +1,124 @@
...
+            data-cy="titleInput"
+            placeholder="New title"
+            value={newTitle}
+            onChange={event => setNewTitle(event.target.value)}
+          />
+          {
+            hasTitleError && (
+              <span className="error">Please enter a title</span>
+            )
+          }
```
Comment (Andr1yK) [inline src/App.tsx:86]:
```suggestion
          {hasTitleError && (
            <span className="error">Please enter a title</span>
          )}
```

**Review by Andr1yK — CHANGES_REQUESTED** [review]
